pkg/agent/monitortool/latency_store.go

```diff
@@ -0,0 +1,256 @@
+	corev1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/util/sets"
+	"k8s.io/klog/v2"
+
+	"antrea.io/antrea/pkg/util/k8s"
+)
+
+// LatencyStore is a store for latency information of connections between Nodes.
+type LatencyStore struct {
+	// Lock for the latency store
+	mutex sync.RWMutex
+
+	// Whether the agent is running in networkPolicyOnly mode
+	isNetworkPolicyOnly bool
+	// The map of Node IP to latency entry, it will be changed by latency monitor
+	nodeIPLatencyMap map[string]*NodeIPLatencyEntry
+	// The map of Node name to Node IP(s), it will be changed by Node watcher
+	// If the agent is running in networkPolicyOnly mode, the value will be the transport IP of the Node.
+	// Otherwise, the value will be the gateway IP of the Node
+	nodeTargetIPsMap map[string][]net.IP
+}
+
+// NodeIPLatencyEntry is the entry of the latency map.
+type NodeIPLatencyEntry struct {
+	// The timestamp of the last sent packet
+	LastSendTime time.Time
+	// The timestamp of the last received packet
+	LastRecvTime time.Time
+	// The last valid rtt of the connection
+	LastMeasuredRTT time.Duration
+}
+
+// NewLatencyStore creates a new LatencyStore.
+func NewLatencyStore(isNetworkPolicyOnly bool) *LatencyStore {
+	store := &LatencyStore{
+		nodeIPLatencyMap:    make(map[string]*NodeIPLatencyEntry),
+		nodeTargetIPsMap:    make(map[string][]net.IP),
+		isNetworkPolicyOnly: isNetworkPolicyOnly,
+	}
+
+	return store
+}
+
+// getNodeIPLatencyEntry returns the NodeIPLatencyEntry for the given Node IP
+// For now, it is only used for testing purposes.
+func (s *LatencyStore) getNodeIPLatencyEntry(nodeIP string) (NodeIPLatencyEntry, bool) {
+	s.mutex.RLock()
+	defer s.mutex.RUnlock()
+
+	entry, ok := s.nodeIPLatencyMap[nodeIP]
+	if !ok {
+		return NodeIPLatencyEntry{}, ok
+	}
+
+	return *entry, ok
+}
+
+// DeleteNodeIPLatencyEntry deletes the NodeIPLatencyEntry for the given Node IP
+func (s *LatencyStore) DeleteNodeIPLatencyEntry(nodeIP string) {
+	s.mutex.Lock()
+	defer s.mutex.Unlock()
+
+	delete(s.nodeIPLatencyMap, nodeIP)
+}
+
+// SetNodeIPLatencyEntry sets the NodeIPLatencyEntry for the given Node IP
+func (s *LatencyStore) SetNodeIPLatencyEntry(nodeIP string, mutator func(entry *NodeIPLatencyEntry)) {
+	s.mutex.Lock()
+	defer s.mutex.Unlock()
+
+	entry, ok := s.nodeIPLatencyMap[nodeIP]
+	if !ok {
+		entry = &NodeIPLatencyEntry{}
+		s.nodeIPLatencyMap[nodeIP] = entry
+	}
+
+	mutator(entry)
+}
+
+// addNode adds a Node to the latency store
+func (s *LatencyStore) addNode(node *corev1.Node) {
+	s.mutex.Lock()
+	defer s.mutex.Unlock()
+
+	s.updateNodeMap(node)
+}
+
+// deleteNode deletes a Node from the latency store
+func (s *LatencyStore) deleteNode(node *corev1.Node) {
+	s.mutex.Lock()
+	defer s.mutex.Unlock()
+
+	nodeIPs, ok := s.nodeTargetIPsMap[node.Name]
+	if !ok {
+		klog.ErrorS(nil, "Failed to get IPs for Node", "nodeName", node.Name)
+		return
+	}
+
+	for _, ip := range nodeIPs {
+		delete(s.nodeIPLatencyMap, ip.String())
+	}
```

Comment:
We are introducing `DeleteStaleNodeIPs` to avoid modifying `nodeIPLatencyMap` from Node event handlers, so this should be removed. I think you have misunderstood the reason why we have to introduce `DeleteStaleNodeIPs`.

Reply:
OK, I have removed it.